Merge the compositing theory sections of Chapter 13 (#506)
browserengineering/book: book/animations.md

:::

Compositing
===========

So, how do we do less work in the raster and draw phase? The answer is
a technique called *compositing*, which just means caching some
rastered images on the GPU and reusing them during later
frames.[^compositing-def]

[^compositing-def]: The term [*compositing*][compositing] just means
combining multiple images together into a final output. In browsers,
we're typically combining rastered images into the final image of the
page, but a similar technique is used in many operating systems to
combine the contents of multiple windows. "Compositing" can also refer
to multi-threaded rendering.

[compositing]: https://en.wikipedia.org/wiki/Compositing

To explain compositing, we'll need to think about our browser's

> display tree. Printing it is going to require similar code in every

Reviewer: ```suggestion
display list. Printing it is going to require similar code in every
```

Reviewer: It is a tree (a forest actually), but we call it a display list anyway throughout.

class DrawRect(DisplayItem):
    def __repr__(self):
        return ("DrawRect(top={} left={} " +
            "bottom={} right={} color={})").format(
            self.left, self.top, self.right,
            self.bottom, self.color)
```

Some of our display commands have a flag to do nothing, like
`ClipRRect`'s `should_clip` flag. It's useful to explicitly indicate
that:

``` {.python}
class ClipRRect(DisplayItem):
    def __repr__(self):
        if self.should_clip:
            return "ClipRRect({})".format(str(self.rrect))
        else:
            return "ClipRRect(<no-op>)"
```

Now we can print out our browser's display list:

``` {.python expected=False}
class Tab:
    def render(self):
        # ...
        for item in self.display_list:
            print_tree(item)
```

For our opacity example, the display list looks like this:

    SaveLayer(alpha=0.112375)
      DrawText(text=This)
      DrawText(text=text)
      DrawText(text=fades)

On the next frame, it instead looks like this:

    SaveLayer(alpha=0.119866666667)
      DrawText(text=This)
      DrawText(text=text)
      DrawText(text=fades)

In each case, rastering this display list means first drawing the

> three words to a surface, and then copy that to the root surface while

Reviewer: ```suggestion
three words to a Skia surface created by `saveLayer`, and then copying that to the root surface while
```

we'll call a *composited layer*:

    Composited Layer 1:
      DrawText(text=This)
      DrawText(text=text)
      DrawText(text=fades)

Now instead of drawing those three words, we can just copy over the
layer:

    SaveLayer(alpha=0.112375)
      DrawCompositedLayer()

Importantly, on the next frame, the `SaveLayer` changes but the
`DrawText`s don't, so on the next frame all we need to do is rerun the
`SaveLayer`:

    SaveLayer(alpha=0.119866666667)
      DrawCompositedLayer()

In other words, the idea behind compositing is to split the display
list into two pieces: a set of composited layers, which are rastered
during the browser's raster phase and then cached, and a *draw display
list*, which is drawn during the browser's draw phase and which uses
the composited layers.

Compositing helps when different frames of the animation have the same
composited layers, which can then be reused, and only change the draw
display list. That's the case here, because the only difference
between frames is the `SaveLayer`, which is in the draw display list.
Now, a browser can choose what composited layers to create however it

> wants. Typically effects like opacity are very fast to execute on a

Reviewer: ```suggestion
wants. Typically visual effects like opacity are very fast to execute on a
```

GPU, but *paint commands* that draw shapes---in our browser,
`DrawText`, `DrawRect`, `DrawRRect`, and `DrawLine`---can be slower.

> Since it's the effects that are typically animated, this means

Reviewer: ```suggestion
Since it's the visual effects that are typically animated, this means
```

> browsers usually leave effects in the draw display list and move paint

Reviewer: ```suggestion
browsers usually leave animated visual effects in the draw display list and move paint
```

commands into composited layers. Of course, in a real browser,

> hardware capabilities, GPU memory, and application data all plays into

Reviewer: ```suggestion
hardware capabilities, GPU memory, and application data all play into
```

> this, but the basic idea of compositing is the same no matter what

Reviewer: ```suggestion
these decisions, but the basic idea of compositing is the same no matter what
```

goes where.

Some animations can't be composited because they affect more than just
the display tree. For example, imagine we animate the `width` of the
`div` above, instead of animating its opacity. In that case, different
frames have different *layout trees*, not just display trees. That
totally changes the coordinates for the `DrawText` calls, and we
wouldn't necessarily be able to reuse the composited layer. Such
animations are called *layout-inducing* and speeding them up requires

> [different techniques](reflow.md).[^not-advisable]

Reviewer: An example  that works in real browsers is here (link to example13-width-transition.html).

Add an exercise to implement `width` animations such that that example works.

Author: Great call.

Reviewer: Is this done? Don't see it yet.

Author: Done now. I got into a nasty git situation.
